book/analysis/performance-of-python-types.md

```diff
@@ -5,90 +5,74 @@ position: 4
-for the array, much like index assignment. However, the
-concatenation operator is $$O(k)$$ where $$k$$ is the size of the list that
-is being concatenated, effectively because $$k$$ sequential assignment
-operations must occur, equivalent to $$k$$ appends.
-
-Popping from a list in Python is typically performed from its end, but
-by passing an index (such as `0`) you can pop from a specific location.
-The performance characteristics of the two are very different. When
-`pop` is called on the end of the list it takes $$O(1)$$ but when pop is
-called on the first element in the list or anywhere in the middle it is
-$$O(n)$$. The reason for this lies in how Python chooses to implement
-lists. When an item is taken from the front of the list, in Python’s
-implementation, all the other elements in the list are shifted one
-position closer to the beginning. This may seem silly to you now, but as
-we explore an alternative list implementation in a later section, we
-will see that this inefficiency is a concession that is made in order to
-allow $$O(1)$$ index operations.
-
-Much like popping from the start or middle of a list, inserting at an
-index is $$O(n)$$ in Python, as every subsequent element must be shifted
-one position closer to the end in order to accommodate the new element.
-Deletion unsurprisingly behaves in the same way.
-
-Iteration—somewhat intuitively—is $$O(n)$$. Iterating over $$n$$ items
-requires $$n$$ steps. This also gives us an intuitive understanding of
-why the `in` operator in Python is $$O(n)$$ when operating over a list—
-effectively we must iterate over every element of the list to determine
-if the element in question is contained therein.
-
-The slice operations require some more thought. To access a slice `[a:b]`
-of an array, we must iterate through every element between that at index
-`a` and that at index `b`. As such slice access is $$O(k)$$ where $$k$$
-is the number of elements in the slice. Deleting a slice is $$O(n)$$ for
-the same reason that deleting a single element is—$$n$$ subsequent
-elements must then be shifted toward the beginning of the list.
-
-As mentioned previously, concatenation is $$O(k)$$ where $$k$$ is the
-length of the concatenated list, due to an effective $$k$$ appends.
-It follows that multiplying a list is $$O(nk)$$ as multiplication of
-a $$k$$ length list $$n$$ times will involve $$k(n - 1)$$ appends, so
-in the order of $$nk$$.
-
-Reversing a list in $$O(n)$$ as each element must be repositioned.
-Finally, as previously mentioned [sort in Python](http://svn.python.org/view/python/trunk/Objects/listsort.txt?revision=69846&view=markup) is $$O(n log n)$$, the
-least intuitive of all of these outcomes, and [beyond the scope of this book](https://en.wikipedia.org/wiki/Timsort)
-to demonstrate.
-
-Below we provide a table of performance characteristics of Python list
-operations, for future reference.
+Python designers had many choices to make when they implemented the list data
```

Comment:
Or switch to something along the lines of "The folk responsible for implementing Python's list data type had many choices to make".